File: ios/runanywhere-sdks-latest/sdk/runanywhere-swift/Sources/RunAnywhere/Public/Extensions/Storage/RunAnywhere+Storage.swift
//
//  RunAnywhere+Storage.swift
//  RunAnywhere SDK
//
//  Public API for storage and download operations.
//

import CRACommons
import Foundation

public extension RunAnywhere {
    /// Register a remote model with the in-memory model registry from a
    /// download URL or Hugging Face reference, through the canonical commons
    /// factory (`rac_register_model_from_url_proto`). Commons derives
    /// id/name/format/artifact, resolves `hf.co/org/repo[:quant]` refs (quant
    /// selection, mmproj pairing, sharded GGUF sets, per-file checksums), and
    /// preserves prior download state when a catalog re-seeds on launch.
    @discardableResult
    static func registerModel(
        id: String? = nil,
        name: String,
        url: String,
        framework: InferenceFramework,
        modality: ModelCategory = .language,
        artifactType: RAModelArtifactType? = nil,
        memoryRequirement: Int64? = nil,
        supportsThinking: Bool = false,
        supportsLora: Bool = false
    ) async throws -> RAModelInfo {
        guard isInitialized else {
            throw SDKException(code: .notInitialized, message: "SDK not initialized", category: .internal)
        }

        var request = RARegisterModelFromUrlRequest()
        request.url = url
        request.name = name
        request.framework = framework
        request.category = modality
        if let id {
            request.id = id
        }
        if let memoryRequirement {
            request.memoryRequiredBytes = memoryRequirement
            request.downloadSizeBytes = memoryRequirement
        }
        if modality.requiresContextLength {
            request.contextLength = 2048
        }
        if supportsThinking {
            request.supportsThinking = true
        }
        if supportsLora {
            request.supportsLora = true
        }
        if let artifactType {
            request.artifactType = artifactType
        }

        return try await CppBridge.ModelRegistry.shared.registerFromUrl(request)
    }

    /// Register an archive-packaged model (tar.gz / tar.bz2 / tar.xz / zip)
    /// where the caller needs to specify the on-disk layout (`directoryBased`,
    /// `nestedDirectory`, etc.) the URL-form `registerModel` cannot infer.
    ///
    /// Builds the archive artifact (type + caller-specified structure) inline,
    /// layers on the caller-supplied capability fields, and persists through the
    /// registry's proto save path in a single `save(...)`.
    @discardableResult
    static func registerModel(
        archive url: String,
        structure: RAArchiveStructure,
        id: String? = nil,
        name: String,
        framework: InferenceFramework,
        modality: ModelCategory = .language,
        archiveType: RAArchiveType? = nil,
        memoryRequirement: Int64? = nil,
        supportsThinking: Bool = false,
        supportsLora: Bool = false
    ) async throws -> RAModelInfo {
        guard isInitialized else {
            throw SDKException(code: .notInitialized, message: "SDK not initialized", category: .internal)
        }

        let downloadURL = URL(string: url)

        // Resolve the archive type (caller override → inference from the URL),
        // then build the archive artifact carrying the caller-specified layout
        // structure that the URL alone cannot infer.
        var archiveArtifact = RAArchiveArtifact()
        if let archiveType {
            archiveArtifact.type = archiveType
        } else if let downloadURL, let inferred = ArchiveType.from(url: downloadURL) {
            archiveArtifact.type = inferred
        }
        archiveArtifact.structure = structure

        var model = RAModelInfo.make(
            id: id ?? generatedModelID(from: url, name: name),
            name: name,
            category: modality,
            format: .unspecified,
            framework: framework,
            downloadURL: downloadURL,
            artifact: .archive(archiveArtifact),
            downloadSizeBytes: memoryRequirement,
            contextLength: modality.requiresContextLength ? 2048 : nil,
            supportsThinking: supportsThinking
        )
        if let memoryRequirement {
            model.memoryRequiredBytes = memoryRequirement
        }
        if supportsLora {
            model.supportsLora = true
        }

        try await CppBridge.ModelRegistry.shared.save(model)
        return model
    }

    /// Register a multi-file model (e.g., VLMs with a separate mmproj, MiniLM
    /// embedding with vocab.txt) through the canonical commons factory
    /// (`rac_register_multi_file_model_proto`) — no URL is involved at the
    /// model level because each `RAModelFileDescriptor` carries its own URL.
    @discardableResult
    static func registerModel(
        multiFile descriptors: [RAModelFileDescriptor],
        id: String,
        name: String,
        framework: InferenceFramework,
        modality: ModelCategory = .language,
        memoryRequirement: Int64? = nil,
        contextLength: Int? = nil,
        supportsThinking: Bool = false,
        source: RAModelSource = .remote
    ) async throws -> RAModelInfo {
        guard isInitialized else {
            throw SDKException(code: .notInitialized, message: "SDK not initialized", category: .internal)
        }

        var request = RARegisterMultiFileModelRequest()
        request.id = id
        request.name = name
        request.framework = framework
        request.category = modality
        request.files = descriptors
        request.source = source
        if let memoryRequirement {
            request.memoryRequiredBytes = memoryRequirement
            request.downloadSizeBytes = memoryRequirement
        }
        if let contextLength {
            request.contextLength = Int32(contextLength)
        } else if modality.requiresContextLength {
            request.contextLength = 2048
        }
        if supportsThinking {
            request.supportsThinking = true
        }

        return try await CppBridge.ModelRegistry.shared.registerMultiFile(request)
    }

    /// Download a registered model. Commons owns planning, transfer (via the
    /// URLSession HTTP adapter), extraction, and validation; Swift owns the
    /// plan → start → poll → import orchestration loop and surfaces the
    /// generated proto progress events to the caller.
    @discardableResult
    static func downloadModel(
        _ model: RAModelInfo,
        onProgress: ((RADownloadProgress) async -> Void)? = nil
    ) async throws -> RADownloadProgress {
        guard isInitialized else {
            throw SDKException(code: .notInitialized, message: "SDK not initialized", category: .network)
        }
        try await ensureServicesReady()

        let resolvedModel = await resolveModelForDownload(model)
        SDKLogger.download.info("Planning download for \(resolvedModel.id)")

        var planRequest = RADownloadPlanRequest()
        planRequest.modelID = resolvedModel.id
        planRequest.model = resolvedModel
        planRequest.resumeExisting = true
        planRequest.validateExistingBytes = true
        planRequest.verifyChecksums = !resolvedModel.checksumSha256.isEmpty

        let plan = await planDownload(planRequest)
        guard plan.canStart else {
            let message = plan.errorMessage.isEmpty ? "Unable to create a download plan" : plan.errorMessage
            SDKLogger.download.error("Download plan rejected for \(resolvedModel.id): \(message)")
            throw SDKException(
                code: .downloadFailed,
                message: message,
                category: .network
            )
        }

        var startRequest = RADownloadStartRequest()
        startRequest.modelID = resolvedModel.id
        startRequest.plan = plan
        startRequest.resume = plan.canResume
        startRequest.resumeToken = plan.resumeToken
        // Commons owns the completion registry mutation: the orchestrator's
        // self-heal calls rac_model_registry_update_download_status, which
        // also persists the durable .rac-manifest.binpb sidecar that restores
        // the entry on the next cold launch.
        startRequest.updateRegistryOnCompletion = true

        let startResult = await CppBridge.Download.shared.start(startRequest)
        guard startResult.accepted else {
            let message = startResult.errorMessage.isEmpty
                ? "The download could not be started"
                : startResult.errorMessage
            SDKLogger.download.error("Download start rejected for \(resolvedModel.id): \(message)")
            throw SDKException(
                code: .downloadFailed,
                message: message,
                category: .network
            )
        }

        SDKLogger.download.info(
            "Download accepted for \(resolvedModel.id) (task=\(startResult.taskID))"
        )

        if startResult.hasInitialProgress {
            let progress = startResult.initialProgress
            if try await reportDownloadProgress(progress, onProgress: onProgress) {
                return progress
            }
        }

        var subscribeRequest = RADownloadSubscribeRequest()
        subscribeRequest.modelID = startResult.modelID.isEmpty ? resolvedModel.id : startResult.modelID
        subscribeRequest.taskID = startResult.taskID

        // Swift owns the polling loop, so a Swift task cancellation must also
        // tear down the native download worker — otherwise the commons
        // download keeps running after the caller's Task ends.
        do {
            while true {
                try Task.checkCancellation()
                try await Task.sleep(nanoseconds: 250_000_000)

                let progress = await CppBridge.Download.shared.pollProgress(subscribeRequest)
                if try await reportDownloadProgress(progress, onProgress: onProgress) {
                    return progress
                }
            }
        } catch is CancellationError {
            await cancelNativeDownload(taskID: subscribeRequest.taskID, modelID: subscribeRequest.modelID)
            throw CancellationError()
        }
    }

    /// Stream download progress for a registered model. Mirrors Kotlin's
    /// `downloadModelStream(_:)` convenience over the callback-based
    /// `downloadModel(_:onProgress:)` API.
    static func downloadModelStream(_ model: RAModelInfo) -> AsyncThrowingStream<RADownloadProgress, Error> {
        AsyncThrowingStream { continuation in
            let task = Task {
                do {
                    _ = try await downloadModel(model) { progress in
                        continuation.yield(progress)
                    }
                    continuation.finish()
                } catch {
                    continuation.finish(throwing: error)
                }
            }
            continuation.onTermination = { termination in
                if case .cancelled = termination {
                    task.cancel()
                }
            }
        }
    }

    /// Import a stable, platform-normalized local model path into the generated
    /// registry. This is also the public local-import entry point for file
    /// picker/bookmark flows after Swift has handled sandbox access.
    static func importModel(_ request: RAModelImportRequest) async throws -> RAModelImportResult {
        guard isInitialized else {
            throw SDKException(code: .notInitialized, message: "SDK not initialized", category: .internal)
        }
        return try await CppBridge.ModelRegistry.shared.importModel(request)
    }

    /// Get storage information as the canonical generated proto result.
    static func getStorageInfo(_ request: RAStorageInfoRequest = RAStorageInfoRequest()) async -> RAStorageInfoResult {
        await CppBridge.Storage.shared.info(request)
    }

    /// Execute or dry-run storage deletion as canonical generated proto data.
    static func deleteStorage(_ request: RAStorageDeleteRequest) async -> RAStorageDeleteResult {
        await CppBridge.Storage.shared.delete(request)
    }

    /// Delete one downloaded model end-to-end: unload it if loaded, remove its
    /// files through the platform adapter, and clear its registry path so the
    /// entry returns to registered-not-downloaded (re-downloadable).
    /// Convenience over `deleteStorage(_:)` with the canonical flag set.
    @discardableResult
    static func deleteModel(_ modelId: String) async -> RAStorageDeleteResult {
        var request = RAStorageDeleteRequest()
        request.modelIds = [modelId]
        request.deleteFiles = true
        request.clearRegistryPaths_p = true
        request.unloadIfLoaded = true
        request.allowPlatformDelete = true
        return await deleteStorage(request)
    }

    /// Clear the SDK's Cache directory. Forwards to `CppBridge.FileManager.clearCache()`,
    /// matching Kotlin's top-level `RunAnywhere.clearCache()` entry point.
    static func clearCache() async throws {
        guard isInitialized else {
            throw SDKException(code: .notInitialized, message: "SDK not initialized", category: .internal)
        }
        try await ensureServicesReady()
        guard CppBridge.FileManager.clearCache() else {
            throw SDKException(code: .deleteFailed, message: "Failed to clear cache", category: .io)
        }
    }

    /// Clear the SDK's Temp directory. Forwards to `CppBridge.FileManager.clearTemp()`,
    /// matching Kotlin's top-level `RunAnywhere.cleanTempFiles()` entry point.
    static func cleanTempFiles() async throws {
        guard isInitialized else {
            throw SDKException(code: .notInitialized, message: "SDK not initialized", category: .internal)
        }
        try await ensureServicesReady()
        guard CppBridge.FileManager.clearTemp() else {
            throw SDKException(code: .deleteFailed, message: "Failed to clean temp files", category: .io)
        }
    }
}

private extension RunAnywhere {
    /// Prefer the registry's canonical metadata when the caller passes a list-row
    /// snapshot that may be missing download_url, archive layout, or checksum fields.
    /// Mirrors Kotlin TC-07 which re-fetches `RAModelInfo` from `listModels()` before
    /// `downloadModel(...)`.
    static func resolveModelForDownload(_ model: RAModelInfo) async -> RAModelInfo {
        var request = RAModelGetRequest()
        request.modelID = model.id
        let getResult = await getModel(request)
        if getResult.found {
            let registryModel = getResult.model
            if !registryModel.downloadURL.isEmpty || model.downloadURL.isEmpty {
                return registryModel
            }
            return model
        }

        let listResult = await listModels()
        guard listResult.success else { return model }
        if let listed = listResult.models.models.first(where: { $0.id == model.id }) {
            if !listed.downloadURL.isEmpty || model.downloadURL.isEmpty {
                return listed
            }
        }
        return model
    }

    /// Plan a download. Oversize-partial self-healing happens inside the
    /// commons planner (validate_existing_bytes deletes stale partials and
    /// replans as a fresh download), so no Swift-side retry loop is needed.
    static func planDownload(_ request: RADownloadPlanRequest) async -> RADownloadPlanResult {
        await CppBridge.Download.shared.plan(request)
    }

    /// Derive a stable model id from a download URL via commons
    /// `rac_model_id_from_url`. Used by the archive overload, whose
    /// caller-specified layout structure the from-url factory cannot express.
    static func generatedModelID(from url: String, name: String) -> String {
        var buffer = [CChar](repeating: 0, count: 256)
        let status = url.withCString { urlPtr in
            rac_model_id_from_url(urlPtr, &buffer, buffer.count)
        }
        if status == RAC_SUCCESS {
            let bytes = buffer.prefix { $0 != 0 }.map { UInt8(bitPattern: $0) }
            let derived = String(bytes: bytes, encoding: .utf8) ?? ""
            if !derived.isEmpty { return derived }
        }
        return name
    }

    /// Tear down the commons download worker for `taskID` / `modelID` so a
    /// Swift `Task.cancel()` propagates through `rac_download_cancel_proto`.
    /// `deletePartialBytes: false` preserves resume tokens for callers that
    /// retry the same model after cancelling.
    static func cancelNativeDownload(taskID: String, modelID: String) async {
        guard !taskID.isEmpty else { return }
        var cancelRequest = RADownloadCancelRequest()
        cancelRequest.taskID = taskID
        cancelRequest.modelID = modelID
        cancelRequest.deletePartialBytes = false
        _ = await CppBridge.Download.shared.cancel(cancelRequest)
    }

    static func reportDownloadProgress(
        _ progress: RADownloadProgress,
        onProgress: ((RADownloadProgress) async -> Void)?
    ) async throws -> Bool {
        if let onProgress {
            await onProgress(progress)
        }

        switch progress.state {
        case .completed:
            return true
        case .failed:
            throw SDKException(
                code: .downloadFailed,
                message: progress.errorMessage.isEmpty ? "Download failed" : progress.errorMessage,
                category: .network
            )
        case .cancelled:
            throw SDKException(code: .cancelled, message: "Download cancelled", category: .network)
        default:
            return progress.stage == .completed
        }
    }

}

File: ios/runanywhere-sdks-latest/sdk/runanywhere-react-native/packages/core/src/Public/Extensions/Storage/RunAnywhere+Storage.ts
/**
 * RunAnywhere+Storage.ts
 *
 * Storage management extension.
 * Delegates to C++ via native module for storage info (C++ handles recursive traversal).
 *
 * Reference: sdk/runanywhere-swift/Sources/RunAnywhere/Public/Extensions/Storage/RunAnywhere+Storage.swift
 */

import { SDKLogger } from '../../../Foundation/Logging/Logger/SDKLogger';
import { requireNativeModule, isNativeModuleAvailable } from '../../../native';
import { ensureServicesReady } from '../../../Foundation/Initialization/ServicesReadyGuard';
import {
  StorageDeleteRequest,
  StorageDeleteResult as StorageDeleteResultCodec,
  StorageInfoRequest,
  StorageInfoResult as StorageInfoResultCodec,
} from '@runanywhere/proto-ts/storage_types';
import type {
  StorageInfo,
  StorageDeleteResult,
  StorageInfoResult,
} from '@runanywhere/proto-ts/storage_types';
import { arrayBufferToBytes } from '../../../services/ProtoBytes';
import { encodeProtoMessage } from '../../../services/ProtoWire';

const logger = new SDKLogger('RunAnywhere.Storage');

export type {
  StorageDeleteCandidate,
  StorageDeleteRequest,
  StorageDeleteResult,
  StorageInfo,
  StorageInfoRequest,
  StorageInfoResult,
} from '@runanywhere/proto-ts/storage_types';

function encode<T>(
  message: T,
  codec: { encode(value: T, writer?: { finish(): Uint8Array }): { finish(): Uint8Array } }
): ArrayBuffer {
  return encodeProtoMessage(message, codec);
}

function decode<T>(
  buffer: ArrayBuffer,
  codec: { decode(bytes: Uint8Array): T },
  fallback: T
): T {
  const bytes = arrayBufferToBytes(buffer);
  return bytes.byteLength === 0 ? fallback : codec.decode(bytes);
}


/**
 * Get canonical generated storage info from native commons.
 * Internal proto plumbing for [getStorageInfo] — Swift has no public
 * `getStorageInfoProto` counterpart, so this stays module-private.
 */
async function getStorageInfoProto(
  request: StorageInfoRequest = {
    includeDevice: true,
    includeApp: true,
    includeModels: true,
    includeCache: true,
  }
): Promise<StorageInfoResult> {
  if (!isNativeModuleAvailable()) {
    return StorageInfoResultCodec.fromPartial({
      success: false,
      errorMessage: 'Native module not available',
    });
  }

  const native = requireNativeModule();
  const buffer = await native.storageInfoProto(
    encode(request, StorageInfoRequest)
  );
  return decode(
    buffer,
    StorageInfoResultCodec,
    StorageInfoResultCodec.fromPartial({
      success: false,
      errorMessage: 'storageInfoProto returned an empty result',
    })
  );
}

/**
 * Execute or dry-run native storage deletion.
 */
export async function deleteStorage(
  request: StorageDeleteRequest
): Promise<StorageDeleteResult> {
  if (!isNativeModuleAvailable()) {
    return StorageDeleteResultCodec.fromPartial({
      success: false,
      errorMessage: 'Native module not available',
    });
  }

  const native = requireNativeModule();
  const buffer = await native.storageDeleteProto(
    encode(request, StorageDeleteRequest)
  );
  return decode(
    buffer,
    StorageDeleteResultCodec,
    StorageDeleteResultCodec.fromPartial({
      success: false,
      errorMessage: 'storageDeleteProto returned an empty result',
    })
  );
}

/**
 * Get generated storage information.
 * Delegates to C++ FileManagerBridge for recursive directory traversal.
 */
export async function getStorageInfo(): Promise<StorageInfo | null> {
  try {
    const result = await getStorageInfoProto();
    return result.success ? result.info ?? null : null;
  } catch (error) {
    logger.warning('Failed to get storage info:', { error });
    return null;
  }
}

/**
 * Clear the SDK's Cache directory. Mirrors Swift `clearCache()` →
 * `CppBridge.FileManager.clearCache()` (cache only — temp files are cleaned
 * by [cleanTempFiles]).
 */
export async function clearCache(): Promise<void> {
  // Clear file caches via native module (C++ handles directory clearing)
  if (isNativeModuleAvailable()) {
    try {
      // Swift parity: RunAnywhere+Storage.swift:309 gates on ensureServicesReady.
      await ensureServicesReady();
      const native = requireNativeModule();
      await native.clearCache();
    } catch (error) {
      logger.warning('Failed to clear native cache:', { error });
    }
  }

  logger.info('Cache cleared');
}

/**
 * Delete one downloaded model end-to-end: unload it if loaded, remove its
 * files through the platform adapter, and clear its registry path so the
 * entry returns to registered-not-downloaded (re-downloadable). Convenience
 * over `deleteStorage` with the canonical flag set — mirrors Swift
 * `RunAnywhere.deleteModel(_:)`.
 */
export async function deleteModel(
  modelId: string
): Promise<StorageDeleteResult> {
  return deleteStorage(
    StorageDeleteRequest.fromPartial({
      modelIds: [modelId],
      deleteFiles: true,
      clearRegistryPaths: true,
      unloadIfLoaded: true,
      allowPlatformDelete: true,
    })
  );
}

/**
 * Clear the SDK's Temp directory. Mirrors Swift `cleanTempFiles()` →
 * `CppBridge.FileManager.clearTemp()`.
 */
export async function cleanTempFiles(): Promise<boolean> {
  if (!isNativeModuleAvailable()) {
    return false;
  }
  try {
    // Swift parity: RunAnywhere+Storage.swift:321 gates on ensureServicesReady.
    await ensureServicesReady();
    const native = requireNativeModule();
    return await native.cleanTempFiles();
  } catch (error) {
    logger.warning('Failed to clean temp files:', { error });
    return false;
  }
}
